In [ ]:
import pandas as pd
import tensorflow as tf
import re
import numpy as np

In [ ]:
#sp = pd.read_csv("spacetime.txt", delimiter = '\t')

In [ ]:
#sp

Spacetime interval
0                           See also: Causal structure
1                   In three dimensions, the distance 
2    Although two viewers may measure the x, y, and...
3    In special relativity, however, the distance b...
4    In four-dimensional spacetime, the analog to d...
..                                                 ...
188                                                  .
189                 {\displaystyle \tan(\theta )=v/c.}
190  The x′ axis is also tilted with respect to the...
191  The same events P, Q, R are plotted in Fig. 2-...
192  While the rest frame has space and time axes t...

[193 rows x 1 columns]

In [ ]:
#delimiter = '\n'
with open('spacetime.txt', 'r') as file:
  spt = file.read()

In [ ]:
spt

'Spacetime interval\nSee also: Causal structure\nIn three dimensions, the distance \n\nAlthough two viewers may measure the x, y, and z position of the two points using different coordinate systems, the distance between the points will be the same for both (assuming that they are measuring using the same units). The distance is "invariant".\n\nIn special relativity, however, the distance between two points is no longer the same if measured by two different observers when one of the observers is moving, because of Lorentz contraction. The situation is even more complicated if the two points are separated in time as well as in space. For example, if one observer sees two events occur at the same place, but at different times, a person moving with respect to the first observer will see the two events occurring at different places, because (from their point of view) they are stationary, and the position of the event is receding or approaching. Thus, a different measure must be used to meas

In [ ]:
spt_data = []
#for i in range(len(spt)):
ssp = re.sub('[^.a-zA-Z]', ' ', spt)
ssp = re.sub('\s+', ' ', ssp)   #removing unnecessary spaces
  #ssp = re.sub('\n\s*\n', '.', ssp)
spt_data.append(str(ssp))
spt_data

['Spacetime interval See also Causal structure In three dimensions the distance Although two viewers may measure the x y and z position of the two points using different coordinate systems the distance between the points will be the same for both assuming that they are measuring using the same units . The distance is invariant . In special relativity however the distance between two points is no longer the same if measured by two different observers when one of the observers is moving because of Lorentz contraction. The situation is even more complicated if the two points are separated in time as well as in space. For example if one observer sees two events occur at the same place but at different times a person moving with respect to the first observer will see the two events occurring at different places because from their point of view they are stationary and the position of the event is receding or approaching. Thus a different measure must be used to measure the effective distance

In [ ]:
sp_new = spt_data[0].split(' ')
sp_new

['Spacetime',
 'interval',
 'See',
 'also',
 'Causal',
 'structure',
 'In',
 'three',
 'dimensions',
 'the',
 'distance',
 'Although',
 'two',
 'viewers',
 'may',
 'measure',
 'the',
 'x',
 'y',
 'and',
 'z',
 'position',
 'of',
 'the',
 'two',
 'points',
 'using',
 'different',
 'coordinate',
 'systems',
 'the',
 'distance',
 'between',
 'the',
 'points',
 'will',
 'be',
 'the',
 'same',
 'for',
 'both',
 'assuming',
 'that',
 'they',
 'are',
 'measuring',
 'using',
 'the',
 'same',
 'units',
 '.',
 'The',
 'distance',
 'is',
 'invariant',
 '.',
 'In',
 'special',
 'relativity',
 'however',
 'the',
 'distance',
 'between',
 'two',
 'points',
 'is',
 'no',
 'longer',
 'the',
 'same',
 'if',
 'measured',
 'by',
 'two',
 'different',
 'observers',
 'when',
 'one',
 'of',
 'the',
 'observers',
 'is',
 'moving',
 'because',
 'of',
 'Lorentz',
 'contraction.',
 'The',
 'situation',
 'is',
 'even',
 'more',
 'complicated',
 'if',
 'the',
 'two',
 'points',
 'are',
 'separated',
 'in',
 'time

In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer()
tokenizer.fit_on_texts([sp_new])
tokenizer.word_index

{'the': 1,
 'of': 2,
 'in': 3,
 'a': 4,
 'is': 5,
 'and': 6,
 's': 7,
 'to': 8,
 '.': 9,
 'spacetime': 10,
 'two': 11,
 'are': 12,
 'x': 13,
 'with': 14,
 'time': 15,
 'frame': 16,
 'that': 17,
 'interval': 18,
 'light': 19,
 'displaystyle': 20,
 'distance': 21,
 'events': 22,
 'at': 23,
 'between': 24,
 'same': 25,
 'event': 26,
 'frames': 27,
 'from': 28,
 'by': 29,
 'space': 30,
 'ct': 31,
 'different': 32,
 'as': 33,
 'than': 34,
 'diagram': 35,
 'be': 36,
 'for': 37,
 'observer': 38,
 'because': 39,
 'or': 40,
 'b': 41,
 'ds': 42,
 'reference': 43,
 'fig.': 44,
 'if': 45,
 'moving': 46,
 'any': 47,
 'other': 48,
 'c': 49,
 'measure': 50,
 'y': 51,
 'z': 52,
 'points': 53,
 'coordinate': 54,
 'will': 55,
 'no': 56,
 'observers': 57,
 'one': 58,
 'it': 59,
 'spatial': 60,
 'which': 61,
 't': 62,
 'can': 63,
 'speed': 64,
 'since': 65,
 'have': 66,
 'world': 67,
 'axis': 68,
 'when': 69,
 'separated': 70,
 'not': 71,
 'an': 72,
 'single': 73,
 'always': 74,
 'standard': 75,
 'pulse':

In [ ]:
len(tokenizer.word_index)

478

In [ ]:
for sentence in spt_data[0].split('.'):
  print(tokenizer.texts_to_sequences([sentence])[0])

[10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11, 194, 115, 50, 1, 13, 51, 6, 52, 116, 2, 1, 11, 53, 117, 32, 54, 195, 1, 21, 24, 1, 53, 55, 36, 1, 25, 37, 80, 196, 17, 81, 12, 197, 117, 1, 25, 118]
[1, 21, 5, 82]
[3, 83, 84, 198, 1, 21, 24, 11, 53, 5, 56, 199, 1, 25, 45, 85, 29, 11, 32, 57, 69, 58, 2, 1, 57, 5, 46, 39, 2, 200, 238]
[1, 202, 5, 203, 204, 205, 45, 1, 11, 53, 12, 70, 3, 15, 33, 206, 33, 3, 30]
[37, 207, 45, 58, 38, 120, 11, 22, 208, 23, 1, 25, 121, 86, 23, 32, 122, 4, 209, 46, 14, 87, 8, 1, 123, 38, 55, 114, 1, 11, 22, 210, 23, 32, 211, 39, 28, 88, 212, 2, 213, 81, 12, 214, 6, 1, 116, 2, 1, 26, 5, 215, 40]
[217, 4, 32, 50, 124, 36, 125, 8, 50, 1, 218, 21, 24, 11, 22]
[3, 127, 128, 10, 1, 219, 8, 21, 5, 1, 18]
[79, 15, 220, 3, 33, 4, 221, 222, 59, 5, 223, 224, 34, 1, 60, 193]
[130, 30, 226, 227, 3, 228, 229, 28, 127, 128, 131, 30]
[1, 132, 133, 37, 230, 30, 6, 15, 231, 10, 5, 17, 30, 6, 15, 12, 232, 71, 82, 61, 5, 8, 134, 17, 233, 1, 234, 235, 32, 57, 55, 236, 89, 1

In [ ]:
input_seq = []
for sentence in spt_data[0].split('.'):
  tokenized_sent = tokenizer.texts_to_sequences([sentence])[0]

  for i in range(1, len(tokenized_sent)):
    input_seq.append(tokenized_sent[:i+1])


In [ ]:
input_seq

[[10, 18],
 [10, 18, 114],
 [10, 18, 114, 78],
 [10, 18, 114, 78, 190],
 [10, 18, 114, 78, 190, 191],
 [10, 18, 114, 78, 190, 191, 3],
 [10, 18, 114, 78, 190, 191, 3, 192],
 [10, 18, 114, 78, 190, 191, 3, 192, 193],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11, 194],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11, 194, 115],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11, 194, 115, 50],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11, 194, 115, 50, 1],
 [10, 18, 114, 78, 190, 191, 3, 192, 193, 1, 21, 79, 11, 194, 115, 50, 1, 13],
 [10,
  18,
  114,
  78,
  190,
  191,
  3,
  192,
  193,
  1,
  21,
  79,
  11,
  194,
  115,
  50,
  1,
  13,
  51],
 [10,
  18,
  114,
  78,
  190,
  191,
  3,
  192,
  193,
  1,
  21,
  79,
  11,
  194,
  115,
  50,

In [ ]:
max_len = max([len(x) for x in input_seq])
max_len

63

In [ ]:
from tensorflow.keras.preprocessing.sequence import pad_sequences
padded_input_seq = pad_sequences(input_seq, maxlen = max_len, padding = 'pre')

In [ ]:
padded_input_seq

array([[  0,   0,   0, ...,   0,  10,  18],
       [  0,   0,   0, ...,  10,  18, 114],
       [  0,   0,   0, ...,  18, 114,  78],
       ...,
       [  0,   0,   0, ...,   8, 187, 188],
       [  0,   0,   0, ..., 187, 188, 189],
       [  0,   0,   0, ..., 188, 189,   1]], dtype=int32)

In [ ]:
X = padded_input_seq[:,:-1]

In [ ]:
y = padded_input_seq[:,-1]

In [ ]:
y

array([ 18, 114,  78, ..., 188, 189,   1], dtype=int32)

In [ ]:
X.shape

(1423, 62)

In [ ]:
y.shape

(1423,)

In [ ]:
from tensorflow.keras.utils import to_categorical
y = to_categorical(y, num_classes = 479)   #479 because len(tokenizer.word_index) + 1; +1 because one-hot-encodin starts from 0.

In [ ]:
y

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.]], dtype=float32)

In [ ]:
y.shape

(1423, 479)

Now **Architecture**<br>
Embedding layer -> LSTM -> Dense Layer(Fully Connected layer)<br>
for one sentence, input to **Embedding layer** is (1,63) because 63 words in each sentence and output is (1,63,100) because each word is cnverted into 100 dimensional vector here.<br>
for **LSTM layer**, there will be 63 time steps (because 63 words in each sentence) and each time step will feed a 100 dimensional input word. LSTM will provide a 150 dimesional output(a hyperparameter).<br>
Dense Layer has 479 nodes becasue vocab size is 478 and final output shape is also required to be of 479 dimensions. Hence Dense layer will perform $σ$(WX + b) on all 479 nodes to give a probability vector of size 479. The highest probability index will give us desired word as prediction.

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense

In [ ]:
model = Sequential()
model.add(Embedding(479, 100, input_length = 62))
model.add(LSTM(150))
model.add(Dense(479, activation = 'softmax'))

In [ ]:
model.compile(loss = 'categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

In [ ]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_3 (Embedding)     (None, 62, 100)           47900     
                                                                 
 lstm_3 (LSTM)               (None, 150)               150600    
                                                                 
 dense_3 (Dense)             (None, 479)               72329     
                                                                 
Total params: 270829 (1.03 MB)
Trainable params: 270829 (1.03 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [ ]:
model.fit(X,y, epochs = 80)

Epoch 1/80
45/45 [==============================] - 7s 102ms/step - loss: 5.7332 - accuracy: 0.0583
Epoch 2/80
45/45 [==============================] - 3s 61ms/step - loss: 5.3253 - accuracy: 0.0654
Epoch 3/80
45/45 [==============================] - 1s 32ms/step - loss: 5.2717 - accuracy: 0.0654
Epoch 4/80
45/45 [==============================] - 1s 17ms/step - loss: 5.2305 - accuracy: 0.0654
Epoch 5/80
45/45 [==============================] - 2s 33ms/step - loss: 5.1789 - accuracy: 0.0661
Epoch 6/80
45/45 [==============================] - 1s 18ms/step - loss: 5.0902 - accuracy: 0.0661
Epoch 7/80
45/45 [==============================] - 1s 15ms/step - loss: 4.9613 - accuracy: 0.0970
Epoch 8/80
45/45 [==============================] - 1s 13ms/step - loss: 4.7760 - accuracy: 0.1195
Epoch 9/80
45/45 [==============================] - 1s 14ms/step - loss: 4.5574 - accuracy: 0.1497
Epoch 10/80
45/45 [==============================] - 1s 13ms/step - loss: 4.3444 - accuracy: 0.1813
Epoch 11

In [ ]:
text = ' same events'
#for next word prediction, first tokenize then do padding then do prediction
text_token = tokenizer.texts_to_sequences([text])[0]

In [ ]:
text_token

[25, 22]

In [ ]:
padded_text_token = pad_sequences([text_token], maxlen = 62, padding = 'pre')   #here pay attention that maxlen took 62 but earlier maxlen took max_len i.e. 63 because earlier target y column is not truncated there.
print(padded_text_token)

[[ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0  0  0  0  0  0  0  0  0 25 22]]


In [ ]:
pos = np.argmax(model.predict(padded_text_token))
pos

1/1 [==============================] - 0s 31ms/step


3

In [ ]:
for word, index in tokenizer.word_index.items():
  if index == pos:
    print(word)
    break

in


In [ ]:
for i in range(6):
  token_text = tokenizer.texts_to_sequences([text])[0]
  padded_token_text = pad_sequences([token_text], maxlen = 62, padding = 'pre')
  pos = np.argmax(model.predict(padded_token_text))

  for word, index in tokenizer.word_index.items():
    if pos == index:
      text = text + ' ' + word
      print(text)
      break

1/1 [==============================] - 0s 20ms/step
 same events in
1/1 [==============================] - 0s 20ms/step
 same events in the
1/1 [==============================] - 0s 18ms/step
 same events in the frame
1/1 [==============================] - 0s 18ms/step
 same events in the frame of
1/1 [==============================] - 0s 20ms/step
 same events in the frame of observer
1/1 [==============================] - 0s 18ms/step
 same events in the frame of observer o


If we also check validation accuracy here, it wont be great. So, to improve performance, increase data, play with hyperparameters and use advanced architectures like stacked LSTM, Bidirectional LSTM or Transformers.